MultiVariantWalker implementation

cmnbroad opened the pull request:
MultiVariantWalker implementation, with VariantRecalibrator and ApplyVQSR switched over to the new base class.

GatherVCFs, MergeVCFs, SelectVariants, and possibly SortVCF are all candidates to be changed to MultiVariantWalkers, but are not included here.

Also, this is a branch of the branch that contains the VQSR and ApplyVQSR tests so I could use those tests to verify the MultiVariantWalker. As result there are two commits here, but the tests are already included a separate PR.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+     * be a substantial number of cache hits between cache misses, reducing the number of times we need to
+     * repopulate the cache from disk.
+     */
+    public static final int DEFAULT_QUERY_LOOKAHEAD_BASES = 1000;
```
Can you use `FeatureDataSource.DEFAULT_QUERY_LOOKAHEAD_BASES` directly here?

cmnbroad replied:
Yes. Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+            new HashSet<>(Arrays.asList(
+                    SequenceDictionaryUtils.SequenceDictionaryCompatibility.COMMON_SUBSET,
+                    SequenceDictionaryUtils.SequenceDictionaryCompatibility.SUPERSET,
+                    SequenceDictionaryUtils.SequenceDictionaryCompatibility.NO_COMMON_CONTIGS));
```
This can go away

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+     * @param featureFile file containing Features
+     * @param name logical name for this data source (may be null)
+     */
+    public void addFeatureDataSource(final File featureFile, final String name) {
```
Would requiring that all `FeatureDataSources` be provided at construction time simplify the code at all? If yes, it might be a good change.

cmnbroad replied:
It simplifies things quite a bit. Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+                    SequenceDictionaryUtils.SequenceDictionaryCompatibility.NO_COMMON_CONTIGS));
+
+    /**
+     * Creates a FeatureDataSource backed by the provided File and assigns this data source the specified logical
```
Comments need to be updated with the new class name

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+    /**
+     * SAMSequenceDictionary used for this (aggregate) source, derived from the individual sources.
+     */
+    private SAMSequenceDictionary mergedDictionary;
```
Does the `mergedHeader` contain the `mergedDictionary`? If not, should the dictionary inside the merged header be set to the merged dictionary?

cmnbroad replied:
Yes. This can have the side effect in some cases of emitting a sequence dictionary into the output file that was derived or partially derived from an index, and which contains 0 for sequence length (whereas in the VariantWalker-derived case no dictionary would be emitted since the input header is used). Should we not write such a sequence dictionary out ?

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+        if (mergedDictionary == null && null != getHeader()) {
+            validateAllSequenceDictionaries();
+            final VCFHeader header = getHeader();
+            mergedDictionary = header.getSequenceDictionary();
```
It would be better if at construction time, the merged header were updated with the fully-merged dictionary

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+            if (mergedDictionary == null || mergedDictionary.isEmpty()) {
+                // last resort - try to get a sequence dictionary from one of the feature data sources (which
+                // will create one from the index if necessary)
+                for (FeatureDataSource<VariantContext> fds : featureDataSources) {
```
The dictionary created from the index will only reflect the variants in that particular `FeatureDataSource`, and may not be usable with the variants from the other data sources, so this fallback may not be a good idea.

cmnbroad replied:
Removed.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+
+    /**
+     * Restricts traversals of this data source via {@link #iterator} to only return Features that overlap the provided
+     //* intervals. Calls to {@link #query(SimpleInterval)} and/or {@link # queryAndPrefetch(SimpleInterval)} are not
```
Comment formatting slightly off here

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+     * @param intervals Our next full traversal will return only Features overlapping these intervals
+     */
+    public void setIntervalsForTraversal( final List<SimpleInterval> intervals ) {
+        validateFeatureSources();
```
`validateFeatureSources()` should probably happen at construction time

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+                    new VCFHeader(VCFUtils.smartMergeHeaders(headers, true)) :
+                    headers.get(0);
+        }
+        return mergedHeader;
```
As mentioned above, should be created at construction time.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+    @Override
+    public void close() {
+        validateFeatureSources();
+        featureDataSources.forEach(dataSource -> dataSource.close());
```
Don't you need to close out any merging iterators that were handed out to clients?

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+            final SAMSequenceDictionary targetDictionary,
+            final SAMSequenceRecord targetSequence)
+    {
+        if (targetSequence.getSequenceLength() != sourceSequence.getSequenceLength()) {
```
I think you want to call into the existing `SequenceDictionaryUtils.sequenceRecordsAreEquivalent()` method to do the name and length checks (since it handles some edge cases), and then do the MD5 check on top of it (although checking the MD5s is stricter than we usually are).

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+                    targetDataSourceName,
+                    targetDictionary
+            );
+        } else if (!Objects.equals(targetSequence.getMd5(), sourceSequence.getMd5())) {
```
I think we might want to tolerate the situation where the MD5 is filled in in one record but not the other, and just emit a warning in that case.

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+    /**
+     * Close the iterator currently open over the data sources, if there is one.
+     */
+    private void closeOpenIterationIfNecessary() {
```
I think your `close()` method should be calling `closeOpenIterationIfNecessary()`

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+
+            @Override
+            public boolean hasNext() {
+                return delegateIterator.hasNext();
```
`return delegateIterator != null && delegateIterator.hasNext()`

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+            }
+
+            @Override
+            public VariantContext next() {
```
throw `NoSuchElementException` if `hasNext()` is false on a call to `next()`

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+     * Invalidate the cached VCFHeader and SAMSequenceDictionary whenever the set of backing FeatureDataSources
+     * changed in order to force them to be recreated and cached on demand.
+     */
+    private void invalidateCachedHeaderAndDictionary() {
```
Can be deleted if initialization of these is tied to construction.

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,408 @@
+    /**
+     * Validate the requirement that one or more variant data sources have been added
+     */
+    private void validateFeatureSources() {
```
This can probably also be deleted if adding data sources is done only at construction.

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantWalker.java:
```diff
@@ -0,0 +1,103 @@
+
+/**
+ * A MultiVariantWalker is a tool that processes one variant at a time from multiple sources of variants, with
+ * optional contextual information from a reference, sets of reads, and/or supplementary sources of Features.
```
Mention that records from the multiple sources are returned in position order.

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantWalker.java:
```diff
@@ -0,0 +1,103 @@
+
+    // NOTE: using File rather than FeatureInput<VariantContext> here so that we can keep this driving source
+    //       of variants separate from any other potential sources of Features
+    @Argument(fullName = StandardArgumentDefinitions.VARIANT_LONG_NAME, shortName = StandardArgumentDefinitions.VARIANT_SHORT_NAME, doc = "A VCF file containing variants", common = false, optional = false)
```
`A VCF file` -> `One or more VCF files`

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantWalker.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    protected void initializeDrivingVariants() {
```
`@Override`

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantWalker.java:
```diff
@@ -0,0 +1,103 @@
+                f -> {
+                    FeatureInput<VariantContext> featureInput = new FeatureInput<>(f, "drivingVariantFile");
+                    if (drivingVariantsFeatureInputs.contains(featureInput)) {
+                        //TODO: should we allow duplicates file names? If so we need to change the FeatureInput hash
```
I don't think we should allow duplicate file names.

cmnbroad replied:
Comment removed.

cmnbroad replied:
Also, this policy is enforced at the MultiVariantWalker level. Should we enforce it at the MultiVariantDataSource level as well ? I don't think there is anything inherent that wouldn't work, so it would be somewhat redundant.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/VariantWalker.java:
```diff
@@ -21,56 +17,21 @@
-        }
-        return super.getBestAvailableSequenceDictionary();
-    }
+    protected Spliterator<VariantContext> getSpliteratorForDrivingVariants() { return drivingVariants.spliterator(); }
```
`@Override`

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantWalker.java:
```diff
@@ -0,0 +1,103 @@
+    @Override
+    public boolean requiresFeatures() { return true; }
+
+    protected SAMSequenceDictionary getSequenceDictionaryForDrivingVariants() { return drivingVariants.getSequenceDictionary(); }
```
`@Override`

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerBase.java:
```diff
@@ -0,0 +1,127 @@
+
+    protected abstract Spliterator<VariantContext> getSpliteratorForDrivingVariants();
+
+    public abstract VCFHeader getHeaderForVariants();
```
javadoc for these abstract methods

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerBase.java:
```diff
@@ -0,0 +1,127 @@
+import java.util.stream.StreamSupport;
+
+/**
+ * Base class for variant walkers, which process one variant at a time from one oe more sources of variants,
```
`oe` -> `or`

cmnbroad replied:
Done.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/SequenceDictionaryUtils.java:
```diff
@@ -95,7 +95,12 @@ public static void validateDictionaries( final String name1,
         final boolean checkContigOrdering = false;
 
-        validateDictionaries(name1, dict1, name2, dict2, requireSuperset, checkContigOrdering);
+        final Set<SequenceDictionaryCompatibility> allowedInCompatibilityParams =
```
These changes to `SequenceDictionaryUtils` can be reverted

cmnbroad replied:
Done.

droazen commented on src/test/java/org/broadinstitute/hellbender/engine/MultiVariantDataSourceUnitTest.java:
```diff
@@ -0,0 +1,145 @@
+import java.util.Arrays;
+import java.util.Iterator;
+
+public final class MultiVariantDataSourceUnitTest extends BaseTest {
```
Can you reproduce some of the query test cases from `FeatureDataSourceUnitTest` here?

droazen commented on src/test/java/org/broadinstitute/hellbender/engine/MultiVariantDataSourceUnitTest.java:
```diff
@@ -0,0 +1,145 @@
+            count++;
+        };
+        Assert.assertEquals(count, 26);
+    }
```
Add a test case in which you make assertions about the order in which variants are returned, like in `FeatureDataSourceUnitTest`

droazen commented on src/test/java/org/broadinstitute/hellbender/engine/MultiVariantWalkerIntegrationTest.java:
```diff
@@ -0,0 +1,157 @@
+        tool.instanceMain(args.toArray(new String[args.size()]));
+        Assert.assertEquals(tool.count, expectedCount);
+    }
+
```
Add a test case in which multiple variants from different files are at the same position, and make sure the traversal can handle it.

cmnbroad replied:
Done.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/SequenceDictionaryUtilsUnitTest.java:
```diff
@@ -214,12 +217,21 @@ public void testSequenceDictionaryValidation( final List<SAMSequenceRecord> firs
+        if (!requireSuperset) {
+            allowedIncompatibilities = new HashSet<>(Arrays.asList(
+                    SequenceDictionaryUtils.SequenceDictionaryCompatibility.COMMON_SUBSET,
+                    SequenceDictionaryUtils.SequenceDictionaryCompatibility.SUPERSET)
```
Can be reverted

cmnbroad replied:
Done.

cmnbroad commented:
@droazen A few questions:

1) The drivingVariants (merged) header is the one used to create the output writer. By pushing the merged sequence dictionary into the merged header, we might write an output file that contains 0 for one or more sequence lengths (if the dictionary came from an index). Should we disallow that? If not then we'll need to be smarter about the merge (I think smartMergeHeader randomly chooses a contig if there is more than one with the same name; we want to force it to choose the non-zero one).

2) If an input file has no embedded dictionary header *and* no index, then we have no dictionary at all, but we need one to create the merging iterator.
- if there is only one input we COULD write special code to not use a merging iterator. Or we could just reject the input. The FeatureDataSourceUnitTest code has such a case.
- If there is more than one input and we have at least one dictionary, but others missing, should we just reject ? This is what I'm currently doing - not sure if there is a sane way to order between these.

cmnbroad commented:
The updates based on code review are done, but I need feedback on couple of questions above before I can finalize this.

cmnbroad commented:
Updated per conversation yesterday:
- If a single input is used, no sequence dictionary is required.
- If multiple inputs are used, at least one must have a sequence dictionary.

Other comments:
- We should consider making FeatureDataSource do the same header/dictionary reconciliation that we're doing here
- In theory you should be able to intermix GenomicsDB/file inputs, but there are no tests for that here (separate ticket ?)
- There is currently no protection against writing out a VCF with a header that contains an index-derived dictionary with zero-len sequences
- I added a GATK3 compatibility arg to VariantRecalibrator in this commit since I was using it for perf testing.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/VariantWalkerBase.java:
```diff
@@ -0,0 +1,144 @@
+     * Return the primary sequence dictionary to be used for the driving variants for this tool. The value returned
+     * will usually have been prepared in {@link #initializeDrivingVariants}
+     */
+    protected abstract SAMSequenceDictionary getSequenceDictionaryForDrivingVariants();
```
Shouldn't this always be the same as `getHeaderForVariants().getSequenceDictionary()`?

cmnbroad replied:
Yes, though it took me a minute to convince myself. I'll implement it that way and remove the overrides.

cmnbroad replied:
I take it back. It's always the same for a MultiVariantWalker, because we make it so by ensuring the header contains the merged dictionary. But for a VariantWalker with a single input with no embedded dictionary, but with an accompanying index, the header will not have the dictionary, but FeatureDataSource will return the synthesized one . Maybe we should make FeatureDataSource behave like MultiFeatureDataSource (over the TODO above) ? Also, I’m wondering if it better perf-wise to cache the dictionary once rather than regenerating it every time from the header, given that hg38 apparently has thousands of contigs.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/VariantRecalibrator.java:
```diff
@@ -294,7 +298,14 @@
+    // Temporary argument for validation of GATK4 implementation against GATKs results:
+    @Advanced
+    @Argument(fullName="gatk3Compatibility", shortName="gatk3",
+            doc="Set initial random number generator state for compatibility with GATK3 results.",
```
"compatibility with" -> "an exact match against"

cmnbroad replied:
Done.

droazen commented on src/test/java/org/broadinstitute/hellbender/engine/MultiVariantDataSourceUnitTest.java:
```diff
@@ -0,0 +1,350 @@
+
+        try (final MultiVariantDataSource multiVariantSource =
+                     new MultiVariantDataSource(featureInputs, FeatureDataSource.DEFAULT_QUERY_LOOKAHEAD_BASES)) {
+            multiVariantSource.getSequenceDictionary();
```
No assertion here? (and no expected exception)

cmnbroad replied:
Assertion added.

droazen commented on src/test/java/org/broadinstitute/hellbender/engine/MultiVariantWalkerIntegrationTest.java:
```diff
@@ -0,0 +1,238 @@
+        int count = 0;
+        SimpleInterval locus;
+
+        public void apply(
```
`@Override`

cmnbroad replied:
Done.

droazen commented on src/test/java/org/broadinstitute/hellbender/engine/MultiVariantWalkerIntegrationTest.java:
```diff
@@ -0,0 +1,238 @@
+        String expectedIDOrder[];
+        int count = 0;
+
+        public void apply(
```
`@Override`

cmnbroad replied:
Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,372 @@
+    final private VCFHeader mergedHeader;
+
+    /**
+     * SAMSequenceDictionary used for this (aggregate) source, derived from the individual sources.
```
Yes, it is, so this be removed. Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,372 @@
+        //
+        validateAllSequenceDictionaries();
+        mergedHeader = getMergedHeader();
+        mergedDictionary = getMergedSequenceDictionary(mergedHeader);
```
Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,372 @@
+    }
+
+    /**
+     * Restricts traversals of this data source via {@link #iterator} to only return Features that overlap the provided
```
Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,372 @@
+        // results in undefined behavior
+        closeOpenIterationIfNecessary();
+
+        if (getSequenceDictionary() != null) {
```
Right. Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,372 @@
+        featureDataSources.forEach(dataSource -> dataSource.close());
+    }
+
+    private SAMSequenceDictionary getMergedSequenceDictionary(VCFHeader header) {
```
Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,372 @@
+
+    /**
+     * We want the headers that are used to create the merged header to have the same sequence dictionary
+     * that was used returned from the data source and used during validation (which may or may not be the
```
Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,372 @@
+     * update it to include the actual sequence dictionary.
+     */
+    //TODO: should FeatureDataSource do this itself ? Right now it returns a seq dict that may not match
+    // the header it returns.
```
Comment removed.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantWalker.java:
```diff
@@ -0,0 +1,105 @@
+    private List<FeatureInput<VariantContext>> drivingVariantsFeatureInputs = new ArrayList<>(2);
+
+    @Override
+    public boolean requiresFeatures() { return true; }
```
Removed.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantWalker.java:
```diff
@@ -0,0 +1,105 @@
+                    features.addToFeatureSources(0, featureInput, VariantContext.class);
+                }
+        );
+        drivingVariants = new MultiVariantDataSource(drivingVariantsFeatureInputs, FeatureDataSource.DEFAULT_QUERY_LOOKAHEAD_BASES);
```
Good catch. Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java:
```diff
@@ -0,0 +1,372 @@
+    final private List<FeatureDataSource<VariantContext>> featureDataSources = new ArrayList<FeatureDataSource<VariantContext>>();
+
+    /**
+     * Merged VCF header used for this (aggregate) source, derived from the individual soruces.
```
Done.
